File: JS/BOJ-S1-2468-Safe_zone.js
'use strict';

const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : '../input.txt';
let input = fs
  .readFileSync(filePath)
  .toString()
  .trim()
  .split('\n')
  .map((v) => v.trimEnd());

let output = [];
const solution = (input) => {
  let ret = 1;
  const N = +input[0];

  const dx = [0, 0, -1, 1];
  const dy = [-1, 1, 0, 0];
  const map = input.slice(1).map((v) => v.split(' ').map(Number));
  const visited = Array.from({ length: N }, () => Array(N).fill(0));

  const dfs = (y, x, depth) => {
    visited[y][x] = depth;
    for (let i = 0; i < 4; i++) {
      const ny = y + dy[i];
      const nx = x + dx[i];
      if (ny < 0 || ny >= N || nx < 0 || nx >= N) continue;
      if (visited[ny][nx] === 0 && map[ny][nx] > depth) {
        dfs(ny, nx, depth);
      }
    }
  };

  for (let d = 1; d <= 100; d++) {
    visited.forEach((v) => v.fill(0));
    let cnt = 0;
    for (let i = 0; i < N; i++) {
      for (let j = 0; j < N; j++) {
        if (visited[i][j] === 0 && map[i][j] > d) {
          dfs(i, j, d);
          cnt++;
        }
      }
    }
    ret = Math.max(ret, cnt);
  }

  output.push(ret);
};

solution(input);
fs.writeFileSync('/dev/stdout', output.join('').trimEnd());
// fs.writeFileSync(
//   '../output.txt',
//   output.join('').trimEnd(),
//   (err) => console.error
// );

File: CPP/BOJ-S1-2468-Safe_zone.cpp
#include <bits/stdc++.h>
using namespace std;

int n;
int a[101][101], visited[101][101], ret = 1;
int dy[4] = {-1, 0, 1, 0};
int dx[4] = {0, 1, 0, -1};

void dfs(int y, int x, int d)
{
  visited[y][x] = 1;
  for (int i = 0; i < 4; ++i)
  {
    int ny = y + dy[i];
    int nx = x + dx[i];
    if (ny < 0 || nx < 0 || ny >= n || nx >= n)
      continue;
    if (a[ny][nx] > d && !visited[ny][nx])
    {
      dfs(ny, nx, d);
    }
  }
  return;
}

int main()
{
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NULL);
  // freopen("input.txt", "r", stdin);
  // freopen("output.txt", "w", stdout);

  cin >> n;
  for (int i = 0; i < n; i++)
  {
    for (int j = 0; j < n; j++)
    {
      cin >> a[i][j];
    }
  }

  for (int d = 1; d < 101; ++d)
  {
    fill(&visited[0][0], &visited[0][0] + 101 * 101, 0);
    int cnt = 0;
    for (int i = 0; i < n; i++)
    {
      for (int j = 0; j < n; j++)
      {
        if (a[i][j] > d && !visited[i][j])
        {
          dfs(i, j, d);
          cnt++;
        }
      }
    }
    ret = max(ret, cnt);
  }

  cout << ret << "\n";
  return 0;
}